### ISSUE TITLE
`valid-jsdoc` incorrectly errors when parameter destructuring is used *before a non-destructured parameter* in a function signature

### ISSUE THREAD

**Issue by tbrisbane** [eslint/eslint#8072]
**Tell us about your environment**
NodeJS, macOS Sierra 10.12.3

* **ESLint Version:** 3.15.0
* **Node Version:** 6.9.2
* **npm Version:** 3.10.9

**What parser (default, Babel-ESLint, etc.) are you using?** Default

**Please show your full configuration:**
```javascript
'use strict';

const ESLintConfiguration = {
  root: true,

  parserOptions: {
    ecmaVersion: 6,
  },

  rules: {
    'valid-jsdoc': [
      'error',
      {
        requireReturn: false,
        requireReturnType: false,
        requireParamDescription: false,
        requireReturnDescription: false,
      },
    ],
  },
};

module.exports = ESLintConfiguration;
```

**What did you do? Please include the actual source code causing the issue.**
Linted this file:
```javascript
'use strict';

/**
 * @param {String} str
 * @param {Number} num
 * @param {Boolean} bool
 * @param {Function} func
 * @param {Object} obj
 */
function jsDocTest(str, { num, bool, func }, obj) {
  console.log({
    str,
    num,
    bool,
    func,
    obj,
  });
}

jsDocTest(
  'a string',
  { num: 12, bool: true, func: () => 42 },
  { some: 'property' }
);
```

**What did you expect to happen?**
There to be no linting errors.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/Users/treybrisbane/jsdoctest/index.js
  3:1  error  Expected JSDoc for 'obj' but found 'bool'  valid-jsdoc

✖ 1 problem (1 error, 0 warnings)
```
---
If I change the function signature to...
```javascript
/**
 * @param {String} str
 * @param {Object} obj
 * @param {Number} num
 * @param {Boolean} bool
 * @param {Function} func
 */
function jsDocTest(str, obj, { num, bool, func }) { ... }
```
...or...
```javascript
/**
 * @param {String} str
 * @param {Number} num
 * @param {Boolean} bool
 * @param {Function} func
 * @param {Symbol} symbol
 * @param {Promise.<String>} promise
 */
function jsDocTest(str, { num, bool, func }, { symbol, promise }) { ... }
```
...then ESLint returns *no errors*.

It seems that `valid-jsdoc` is incorrectly handling the case of destructuring appearing *before a non-destructured parameter*.

**Comment by not-an-aardvark** [comment]
I think the issue here is that `valid-jsdoc` just doesn't handle destructuring at all. It assumes that each `@param` tag maps to a single parameter (even a destructured parameter), and it doesn't validate that destructured parameters have `@param` tags.

This seems related to https://github.com/eslint/eslint/issues/2051, and there is also a [TODO comment](https://github.com/eslint/eslint/blob/c7e64f30de7aa56e6e0c00ffd1f9ab398920b0b1/lib/rules/valid-jsdoc.js#L362) in `valid-jsdoc` for handling destructuring in a reasonable way.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
